src/backend/backend.did

```diff
@@ -1,150 +1,175 @@
 };
 
 type ApiStringResult = variant {
-    Ok : text;
-    Err : ApiError;
+  Ok : text;
+  Err : ApiError;
 };
 
 type ApiNatResult = variant {
-    Ok : nat64;
-    Err : ApiError;
+  Ok : nat64;
+  Err : ApiError;
 };
 
 type ApiFloatResult = variant {
-    Ok : float64;
-    Err : ApiError;
+  Ok : float64;
+  Err : ApiError;
 };
 
 type ApiEmptyResult = variant {
-    Ok;
-    Err : ApiError;
+  Ok;
+  Err : ApiError;
 };
 
 type TimestampNs = nat64;
 
 type UserId = principal;
 
 type UserRole = variant {
-    Admin;
-    Deployer;
+  Admin;
+  Deployer;
 };
 
 type User = record {
-    role : UserRole;
-    created_at : TimestampNs;
-    payments : vec nat64;
+  role : UserRole;
+  created_at : TimestampNs;
+  payments : vec nat64;
 };
 
 type GetUserResult = variant {
-    Ok : User;
-    Err : ApiError;
+  Ok : User;
+  Err : ApiError;
 };
 
 type CreateUserResult = variant {
-    Ok : UserId;
-    Err : ApiError;
+  Ok : UserId;
+  Err : ApiError;
 };
 
 type DeploymentId = text;
 
 type DeploymentState = variant {
-    Initialized;
-    DeploymentCreated : record {
-        tx_hash : text;
-        dseq : nat64;
-        manifest_sorted_json : text;
-    };
-    LeaseCreated : record {
-        tx_hash : text;
-        provider_url : text;
-    };
-    Active;
-    Closed;
-    FailedOnClient : record {
-        reason : text;
-    };
-    FailedOnCanister : record {
-        reason : text;
-    };
+  Initialized;
+  DeploymentCreated : record {
+    tx_hash : text;
+    dseq : nat64;
+    manifest_sorted_json : text;
+  };
+  LeaseCreated : record {
+    tx_hash : text;
+    provider_url : text;
+  };
+  Active;
+  Closed;
+  FailedOnClient : record {
+    reason : text;
+  };
+  FailedOnCanister : record {
+    reason : text;
+  };
 };
 
 type Deployment = record {
-    sdl : text;
-    user_id : UserId;
-    state_history : vec record { TimestampNs; DeploymentState };
+  sdl : text;
+  user_id : UserId;
+  state_history : vec record { TimestampNs; DeploymentState };
+  icp_price : float64;
 };
 
 type GetDeploymentResult = variant {
-    Ok : record {
-        id : DeploymentId;
-        deployment : Deployment;
-    };
-    Err : ApiError;
+  Ok : record {
+    id : DeploymentId;
+    deployment : Deployment;
+  };
+  Err : ApiError;
 };
 
 type GetDeploymentsResult = variant {
-    Ok : vec record {
-        id : DeploymentId;
-        deployment : Deployment;
-    };
-    Err : ApiError;
+  Ok : vec record {
+    id : DeploymentId;
+    deployment : Deployment;
+  };
+  Err : ApiError;
 };
 
 type CreateDeploymentResult = variant {
-    Ok : DeploymentId;
-    Err : ApiError;
+  Ok : DeploymentId;
+  Err : ApiError;
+};
+
+type DeploymentParams = record {
+  name : opt text;
+  image : opt text;
+  env_vars : opt vec record { text; text };
+  ports : vec record { nat32; opt nat32; text };
+  cpu : CpuSize;
+  memory : MemoryStorageSize;
+  storage : MemoryStorageSize;
+  volume_mount : opt text;
+  command : opt vec text;
```

Comment:
- `image` and `name` should be required
- `env_vars` and `command` should just be `vec`s since candid threats them as arrays in both cases in javascript

Reply:
If we make image and name non optional, then we have to initialized them to String::new() when creating DeploymentParams. However, this way, if the user does not explicitly set them by calling the methods name() and image(), the try_from_deployment_params() method will not fail even if it is trying to deploy an invalid image.
 But the setter methods are only used for tests so I guess it's fine to remove the opt